cli/internal/cache/cache_http.go

```diff
@@ -206,56 +207,73 @@ func (cache *httpCache) retrieve(hash string) (bool, []string, int, error) {
 			err = fmt.Errorf("artifact verification failed: artifact tag does not match expected tag %s", expectedTag)
 			return false, nil, 0, err
 		}
+		// We read the old artifactReader, we can close it.
+		// The defered call to artifactReader.Close will pick up the new value
+		_ = artifactReader.Close()
 		// The artifact has been verified and the body can be read and untarred
 		artifactReader = ioutil.NopCloser(bytes.NewReader(b))
 	}
+	files, err := restoreTar(cache.repoRoot, artifactReader)
+	if err != nil {
+		return false, nil, 0, err
+	}
+	return true, files, duration, nil
+}
+
+func restoreTar(root fs.AbsolutePath, reader io.Reader) ([]string, error) {
 	files := []string{}
 	missingLinks := []*tar.Header{}
-	gzr, err := gzip.NewReader(artifactReader)
+	gzr, err := gzip.NewReader(reader)
 	if err != nil {
-		return false, nil, 0, err
+		return nil, err
 	}
-	defer gzr.Close()
+	defer func() { _ = gzr.Close() }()
 	tr := tar.NewReader(gzr)
 	for {
 		hdr, err := tr.Next()
 		if err != nil {
 			if err == io.EOF {
 				for _, link := range missingLinks {
-					err := restoreSymlink(link, true)
+					err := restoreSymlink(root, link, true)
 					if err != nil {
-						return false, nil, 0, err
+						return nil, err
 					}
 				}
 
-				return true, files, duration, nil
+				return files, nil
 			}
-			return false, nil, 0, err
+			return nil, err
 		}
 		files = append(files, hdr.Name)
+		filename := root.Join(hdr.Name)
+		if isChild, err := root.ContainsPath(filename); err != nil {
+			return nil, err
+		} else if !isChild {
+			return nil, fmt.Errorf("cannot untar file to %v", filename)
```

Comment:
(We could do that by returning `files` and then clean up at the `restoreTar`  call site.)

Reply:
We can't revert. We can delete files we've written, but we can't restore to previous state because we have potentially truncated existing files. Either way, you're in an inconsistent state and need to start over.

It might be worth considering what to do in this scenario, your best bet is likely engineering a cache miss. If you've gotten into this scenario accidentally, you likely need to make a code change anyways, which will force a miss.